src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,231 @@ namespace mlpack {
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ * 
+ * This sequential bootstrap is suitable for time-series that are not
+ * independent and identically distributed (IID). The bootstrap will
+ * generate variations of the data that are more IID. In order to do
+ * that it takes an indicator matrix that is a measure for how
+ * much informational overlap exists between each data point in the
+ * data set. Then it will randomly draw from the dataset with
+ * replacement but reducing the likelihood to draw data points that
+ * have high informational overlap with already drawn samples.
+ *
+ * @tparam IndMatType Indicator matrix type.
+ */
+template<typename IndMatType = arma::umat>
+class SequentialBootstrap
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param[in] intervals See ComputeAverageUniqueness().
+   */
+  SequentialBootstrap(const IndMatType& intervals) :
+    intervals(intervals)
+  {
+    if (intervals.n_rows != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2 x m matrix!");
+  }
+
+  /**
+   * Compute the average uniqueness of each event at any sampling time point.
+   * The average uniqueness is a measure for how isolated an event is during
+   * its lifetime from other events.
+   *
+   * @param[in] intervals Is a `2 x m` matrix, where each of the m columns has
+   *                      the start sample and the end sample of an interval.
+   *                      This matrix is a space-efficient form of the
+   *                      indicator matrix.
+   * @param[in] indices Indices of the average uniqueness that is returned.
+   *                    This parameter is used for optimization purposes so
+   *                    that in the unit tests the average uniqueness for
+   *                    all events can be accessed while in the production
+   *                    code only a single element is accessed.
+   * @return The average uniqueness of the events in @p intervals.
+   */
+  static arma::vec ComputeAverageUniqueness(
+    const IndMatType& intervals,
+    const arma::uvec& indices,
+    const arma::vec&  concurrency)
+  {
+    arma::vec       avg(indices.n_rows, arma::fill::zeros);
+    const arma::vec invConcurrency(
+      arma::vec(concurrency.n_rows, arma::fill::ones) / concurrency);
```

Comment:
Ok---I now agree that `ComputeAverageUniqueness()` is correct.  Everything makes sense.  However I think it should be possible to maintain `invConcurrency` as well as `concurrency`, so that this function does not have to iterate over the full concurrency vector when a single element is given for `indices`.